README
autobox version 2.60
====================

The autobox pragma allows methods to be called on integers, floats, strings, arrays,
hashes, and code references in exactly the same manner as blessed references.

The autoboxing is transparent: boxed values are not blessed into their (user-defined)
implementation class (unless the method elects to bestow such a blessing) - they simply
use its methods as though they are.

autobox is lexically scoped, and bindings for an outer scope
can be overridden or countermanded in a nested scope.

See the POD for more details.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Scope::Guard

COPYRIGHT AND LICENCE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

Copyright (c) 2003-2010 chocolateboy <chocolate@cpan.org>


Changes
Revision history for Perl extension autobox

2.60  Wed Mar 17 16:34:56 2010
    - fix RT #46814 (thanks Tye McQueen)
      - added t/rt_46814
    - fix RT #49273 (thanks Daniel Austin)
    - fix RT #55565 (thanks Schwern)
    - fix RT #55652 (thanks Schwern)
      - $native->isa and $native->can must now be called as $native->autobox_isa and $native->autobox_can
      - added t/rt_55652

2.55  Sun May 25 03:20:54 2008
    - fix MANIFEST again - restore Changes

2.54  Sun May 25 00:36:04 2008
    - fix MANIFEST

2.53  Sat May 24 22:19:45 2008
    - add support for UNIVERSAL virtual type
        - added t/universal.t
    - moved autobox::type method to autobox::universal::type subroutine
        - added export.t
    - added t/default.t
    - portability fix for non-gcc compilers (thanks chris)
    - misc code/documentation fixes/cleanups

2.52  Tue May 20 12:24:01 2008
    - more type fixes

2.51  Tue May 20 10:40:32 2008
    - fix type identification for former INTEGERs and FLOATs (thanks Mitchell N Charity)
        - added type.t
    - fix for perl 5.11 (thanks Andreas Koenig)
    - document eval EXPR gotcha

2.50  Mon May 19 17:39:22 2008
    - add support for INTEGER, FLOAT, NUMBER and STRING
        - added scalar.t
    - updated documentation

2.43  Thu May 15 21:14:08 2008
    - fix @isa bug
        - added t/isa.t
    - scope cleanup
    - documentation tweak

2.42  Tue May 13 22:22:55 2008
    - upgrade ppport.h to 3.13_03 to s/workaround/fix/

2.41  Tue May 13 20:02:37 2008
    - work around $value->$method segfault with non-string method names under perls <= 5.8.8
    - added license info

2.40  Mon May 12 23:51:26 2008
    - support @array and %hash (thanks Yuval Kogman (nothingmuch) and Matthijs van Duin (xmath))
        - added t/autoref.t
    - fix $value->$method segfault with undef, integer, float &c. (i.e. non-string)
      method names (thanks John Goulah)

2.30  Fri May  9 01:52:19 2008
    - support $value->$method, where $method is a method name or subroutine reference:
        - added t/name.t
        - added t/coderef.t

2.23  Sun Feb 24 15:17:05 2008
    - rm redundant $^H hacking

2.22  Sun Feb 24 14:44:58 2008
    - added hints.t

2.21  Fri Feb 22 21:40:54 2008
    - merge unimport.t and time_travel.t into unmerge.t
    - more tests

2.20  Thu Feb 21 23:30:53 2008
    - Fix broken merging
        - corrected merge.t
        - added time_travel.t to verify correctness

2.11  Wed Feb 20 21:06:25 2008
    - Windows portability fix: ANSIfy C99-ism (thanks Taro Nishino)
    - revert broken micro-optimization

2.10  Wed Feb 20 02:16:42 2008
    - fix + tests:
        unimport
        default namespace(s) in an array ref

2.02  Sun Feb 17 16:59:28 2008
    - doc tweak
    - POD formatting

2.01  Sun Feb 17 03:56:22 2008
    - documentation fix:
        rm reference to $class->SUPER::import(TYPE => __PACKAGE__)
            and explain why an auxiliary class should be used
        

2.00  Sun Feb 17 02:29:11 2008
    - API changes:
        autobox with one or more args leaves the unspecified types unboxed
        multiple autobox (or autobox subclass) invocations in the same lexical scope
            are merged (thanks Matsuno Tokuhiro)
        multiple bindings for each type can be supplied as an ARRAY ref of classes
            or namespaces
        "no autobox qw(...)" disables/resets bindings for the specified type(s)

    - fixed incorrect bareword handling
    - perl 5.10 compatibility fixes (thanks Andreas Koenig)
    - document previously undocumented features
    - document subclassing
    - merge.t: test merging
    - beef up the default DEBUG handler so that it shows the superclasses
      of the synthetic classes
    - Windows compatibilty fix (thanks Alexandr Ciornii)
    - misc optimizations, cleanups

1.22  Sun Sep 23 22:27:44 2007
    - (Perl_ck_subr and Perl_ck_null): fix build failure on Windows

1.21  Sun Sep 23 20:35:37 2007
    - (Makefile): fix build failure on Windows (thanks Alexandr Ciornii)

1.20  Sun Sep 23 14:05:39 2007
    - (ptable.h): fix build failures on perl >= 5.9.3 (thanks Andreas Koenig)
    - (Perl_pp_method_named): fix build failure on Windows (thanks randyk and Alexandr Ciornii)

1.10  Thu Nov 23 20:32:53 2006
    - moved END handler into XS
    - updated SEE ALSO section
    - s/REPORT/DEBUG/
    - fix and test for UNDEF => ''
    - portability fixlet for Windows

1.04  Mon Nov 20 00:25:50 2006
    - fix threaded perl pessimization
    - applied patch: http://rt.cpan.org/Public/Bug/Display.html?id=22868 (thanks Steve Peters)
    - documentation fixlet
    - portability fixlet

1.03  Sat Apr 23 20:35:16 2005
    - workaround and test for %^H bug
    - require perl >= 5.8

1.02  Tue Apr 12 20:52:02 2005
    - re-fixed Makefile.PL/META.yml + copyright

1.01  Tue Apr 12 19:58:49 2005
    - compatibility/portability fixes + isolate ptr table from perl's implementation

1.00  Tue Apr 12 01:16:52 2005
    - rewrite: no longer requires a patch

0.11  Tue Feb 3 13:21:47 2004
    - Added patch for perl-5.8.3

0.10  Fri Dec 12 15:24:16 2003
    - fixed obsolete reference to perl-5.8.1 in POD

0.09  Fri Dec 12 11:53:02 2003
    - Added patch for perl-5.8.2

0.08  Fri Oct 17 11:50:34 2003
    - removed obsolete references to perl-5.8.1-RC4 from README

0.07  Tue Oct 14 13:34:16 2003
    - updated patch to work against perl-5.8.1. This patch should be
      applied to a clean perl-5.8.1 tree. Previous versions of perl are
      no longer supported
    - minor documentation tweaklets
    - added typemap() static method to autobox.pm to facilitate subclassing

0.06  Mon Aug 18 17:40:53 2003

    - This version provides an updated patch. It should be applied to a
      clean perl-5.8.1-RC4 tree
    - Thanks to Tassilo von Parseval for hunting down and fixing a
      memory leak
    - Added support for builtin pseudotype, UNDEF
    - Added tests and documentation for old VERSION() and new UNDEF
      features

0.05  Mon Aug 11 03:13:04 2003


lib/autobox.pm
package autobox;

use 5.008;

use strict;
use warnings;

use Carp;
use XSLoader;
use Scalar::Util;
use Scope::Guard;
use Storable;

our $VERSION = '2.60';

XSLoader::load 'autobox', $VERSION;

use autobox::universal (); # don't import

############################################# PRIVATE ###############################################

my $SEQ            = 0;  # unique identifier for synthetic classes
my $BINDINGS_CACHE = {}; # hold a reference to the bindings hashes
my $CLASS_CACHE    = {}; # reuse the same synthetic class if the @isa has been seen before

# all supported types
# the boolean indicates whether the type is a real internal type (as opposed to a virtual type)
my %TYPES = (
    UNDEF     => 1,
    INTEGER   => 1,
    FLOAT     => 1,
    NUMBER    => 0,
    STRING    => 1,
    SCALAR    => 0,
    ARRAY     => 1,
    HASH      => 1,
    CODE      => 1,
    UNIVERSAL => 0
);

# type hierarchy: keys are parents, values are (depth, children) pairs
my %ISA = (
    UNIVERSAL => [ 0, [ qw(SCALAR ARRAY HASH CODE) ] ],
    SCALAR    => [ 1, [ qw(STRING NUMBER) ] ],
    NUMBER    => [ 2, [ qw(INTEGER FLOAT) ] ]
);

# default bindings when no args are supplied
my %DEFAULT = (
    SCALAR => 'SCALAR',
    ARRAY  => 'ARRAY',
    HASH   => 'HASH',
    CODE   => 'CODE'
);

# reinvent List::MoreUtils::uniq to keep the dependencies light - return a reference
# to an array containing (in order) the unique members of the supplied list
sub _uniq($) {
    my $list = shift;
    my (%seen, @uniq);

    for my $element (@$list) {
        next if ($seen{$element});
        push @uniq, $element;
        $seen{$element} = 1;
    }

    return [ @uniq ];
}

# create a shim class - actual methods are implemented by the classes in its @ISA
#
# as an optimization, return the previously-generated class
# if we've seen the same (canonicalized) @isa before
sub _generate_class($) {
    my $isa = _uniq(shift);

    # As an optimization, simply return the class if there's only one.
    # This speeds up method lookup as the method can (often) be found directly in the stash
    # rather than in the ISA hierarchy with its attendant AUTOLOAD-related overhead
    if (@$isa == 1) {
        my $class = $isa->[0];
        _universalize($class); # nop if it's already been universalized
        return $class;
    }

    my $key = Storable::freeze($isa);

    return $CLASS_CACHE->{$key} ||= do {
        my $class = sprintf('autobox::shim::<%d>', ++$SEQ);
        my $synthetic_class_isa = _get_isa($class); # i.e. autovivify

        @$synthetic_class_isa = @$isa;
        _universalize($class);
        $class;
    };
}

# make can() and isa() work as expected for autoboxed values by calling the method on their associated class
sub _universalize ($) {
    my $class = shift;
    return unless (defined $class);
    {
        no strict 'refs';
	# don't interfere with STRING->can and STRING->isa: https://rt.cpan.org/Ticket/Display.html?id=55565
        *{"$class\::autobox_can"} = sub { shift; $class->can(@_) } unless (*{"$class\::autobox_can"}{CODE});
        *{"$class\::autobox_isa"} = sub { shift; $class->isa(@_) } unless (*{"$class\::autobox_isa"}{CODE});
    }
}

# pretty-print the bindings hash by showing its values as the inherited classes rather than the synthetic class
sub _pretty_print($) {
    my $hash = { %{ shift() } }; # clone the hash to isolate it from the original

    # reverse() turns a hash that maps an isa signature to a class name into a hash that maps
    # a class name into a boolean
    my %synthetic = reverse(%$CLASS_CACHE);

    for my $type (keys %$hash) {
        my $class = $hash->{$type};
        $hash->{$type} = $synthetic{$class} ? [ _get_isa($class) ] : [ $class ];
    }

    return $hash;
}

# default sub called when the DEBUG option is supplied with a true value
# prints the assigned bindings for the current scope
sub _debug ($) {
    my $bindings = shift;
    require Data::Dumper;
    no warnings qw(once);
    local ($|, $Data::Dumper::Indent, $Data::Dumper::Terse, $Data::Dumper::Sortkeys) = (1, 1, 1, 1);
    print STDERR Data::Dumper::Dumper($bindings), $/;
}

# return true if $ref ISA $class - works with non-references, unblessed references and objects
# we can't use UNIVERSAL::isa to test if a value is an array ref;
# if the value is 'ARRAY', and that package exists, then UNIVERSAL::isa('ARRAY', 'ARRAY') is true!
sub _isa($$) {
    my ($ref, $class) = @_;
    return Scalar::Util::blessed($ref) ? $ref->isa($class) : ref($ref) eq $class;
}

# get/autovivify the @ISA for the specified class
sub _get_isa($) {
    my $class = shift;
    my $isa   = do {
        no strict 'refs';
        *{"$class\::ISA"}{ARRAY};
    };
    return wantarray ? @$isa : $isa;
}

# install a new set of bindings for the current scope
#
# XXX this could be refined to reuse the same hashref if its contents have already been seen,
# but that requires each (frozen) hash to be cached; at best, it may not be much of a win, and at
# worst it will increase bloat
sub _install ($) {
    my $bindings = shift;
    $^H{autobox} = $bindings;
    $BINDINGS_CACHE->{$bindings} = $bindings; # keep the $bindings hash alive
}

# return the supplied class name or a new class name made by appending the specified
# type to the namespace prefix
sub _expand_namespace($$) {
    my ($class, $type) = @_;

    # make sure we can weed out classes that are empty strings or undef by returning an empty list
    Carp::confess("_expand_namespace not called in list context") unless (wantarray);

    if ((defined $class) && ($class ne '')) {
        ($class =~ /::$/) ? "$class$type" : $class;
    } else { # return an empty list
        ()
    }
}

############################################# PUBLIC (Methods) ###############################################

# enable some flavour of autoboxing in the current scope
sub import {
    my ($class, %args) = @_;
    my $debug = delete $args{DEBUG};

    %args = %DEFAULT unless (%args); # wait till DEBUG has been deleted

    # normalize %args so that it has a (possibly empty) array ref for all types, both real and virtual
    for my $type (keys %TYPES) {
        if (exists $args{$type}) { # exists() as the value may be undef (or ''), meaning "don't default this type"
            if (_isa($args{$type}, 'ARRAY')) {
                $args{$type} = [ @{$args{$type}} ]; # clone the array ref to isolate changes
            } else {
                $args{$type} = [ $args{$type} ];
            }
        } else {
            $args{$type} = [];
        }
    }

    # if supplied, fill in defaults for unspecified SCALAR, ARRAY, HASH and CODE bindings
    # must be done before the virtual type expansion below as one of the defaults, SCALAR, is a
    # virtual type

    my $default = delete $args{DEFAULT};

    if ($default) {
        $default = [ $default ] unless (_isa($default, 'ARRAY')); # no need to clone as we flatten it each time

        for my $type (keys %DEFAULT) {
            # don't default if a binding has already been supplied; this may include an undef value meaning
            # "don't default this type" e.g.
            #
            #     use autobox
            #         DEFAULT => 'MyDefault',
            #         HASH    => undef;
            #
            # undefs are winnowed out by _expand_namespace

            next if (@{$args{$type}}); 
            push @{$args{$type}}, map { _expand_namespace($_, $type) } @$default;
        }
    }

    # expand the virtual type "macros" from the root to the leaves
    for my $vtype (sort { $ISA{$a}->[0] <=> $ISA{$b}->[0] } keys %ISA) {
        next unless ($args{$vtype});

        my @types = @{$ISA{$vtype}->[1]};

        for my $type (@types) {
            if (_isa($args{$vtype}, 'ARRAY')) {
                push @{$args{$type}}, map { _expand_namespace($_, $vtype) } @{$args{$vtype}};
            } else {
                # _expand_namespace returns an empty list if $args{$vtype} is undef (or '')
                push @{$args{$type}}, _expand_namespace($args{$vtype}, $vtype);
            }
        }

        delete $args{$vtype};
    }

    my $bindings; # custom typemap

    # clone the bindings hash if available
    #
    # we may be assigning to it, and we don't want to contaminate outer/previous bindings
    # with nested/new bindings
    #
    # as of 5.10, references in %^H get stringified at runtime, but we don't need them then

    $bindings = $^H{autobox} ? { %{ $^H{autobox} } } : {};

    # sanity check %args, expand the namespace prefixes into class names,
    # and copy values to the $bindings hash

    my %synthetic = reverse (%$CLASS_CACHE); # synthetic class name => bool - see _pretty_print

    for my $type (keys %args) {
        # we've handled the virtual types, so we only need to check that this is a valid (real) type
        Carp::confess("unrecognized option: '", (defined $type ? $type : '<undef>'), "'") unless ($TYPES{$type});

        my (@isa, $class);
       
        if ($class = $bindings->{$type}) {
            @isa = $synthetic{$class} ? _get_isa($class) : ($class);
        }

        # perform namespace expansion; dups are removed in _generate_class below
        push @isa, map { _expand_namespace($_, $type) } @{$args{$type}};

        $bindings->{$type} = [ @isa ]; # assign the (possibly) new @isa for this type
    }

    # replace each array ref of classes with the name of the generated class.
    # if there's only one class in the type's @ISA (e.g. SCALAR => 'MyScalar') then
    # that class is used; otherwise a shim class whose @ISA contains the two or more classes
    # is created 

    for my $type (keys %$bindings) {
        my $isa = $bindings->{$type};

        # delete empty arrays e.g. use autobox SCALAR => []
        if (@$isa == 0) {
            delete $bindings->{$type};
        } else {
            # associate the synthetic/single class with the specified type
            $bindings->{$type} = _generate_class($isa); 
        }
    }

    # This turns on autoboxing i.e. the method call checker sets a flag on the method call op
    # and replaces its default handler with the autobox implementation.
    #
    # It needs to be set unconditionally because it may have been unset in unimport

    $^H |= 0x120000; # set HINT_LOCALIZE_HH + an unused bit to work around a %^H bug

    # install the specified bindings in the current scope
    _install($bindings);

    # this is %^H as an integer - it changes as scopes are entered/exited
    # we don't need to stack/unstack it in %^H as %^H itself takes care of that
    # note: we need to call this *after* %^H is referenced (and possibly created) above

    my $scope = _scope();
    my $old_scope = exists($^H{autobox_scope})? $^H{autobox_scope} : 0;
    my $new_scope; # is this a new (top-level or nested) scope?

    if ($scope == $old_scope) {
        $new_scope = 0;
    } else {
        $^H{autobox_scope} = $scope;
        $new_scope = 1;
    }

    # warn "OLD ($old_scope) => NEW ($scope): $new_scope ", join(':', (caller(1))[0 .. 2]), $/;

    if ($debug) {
        $debug = \&_debug unless (_isa($debug, 'CODE'));
        $debug->(_pretty_print($bindings));
    }

    return unless ($new_scope);

    # This sub is called when this scope's $^H{autobox_leave} is deleted, usually when
    # %^H is destroyed at the end of the scope, but possibly directly in unimport()
    #
    # _enter splices in the autobox method call checker and method call op
    # if they're not already enabled
    #
    # _leave performs the necessary housekeeping to ensure that the default
    # checker and op are restored when autobox is no longer in scope

    my $guard = Scope::Guard->new(sub { _leave() });
    $^H{autobox_leave} = $guard;

    _enter();
}

# delete one or more bindings; if none remain, disable autobox in the current scope
#
# note: if bindings remain, we need to create a new hash (initially a clone of the current
# hash) so that the previous hash (if any) is not contaminated by new deletions(s)
#
#   use autobox;
#
#       "foo"->bar;
#
#   no autobox qw(SCALAR); # don't clobber the default bindings for "foo"->bar
#
# however, if there are no more bindings we can remove all traces of autobox from the
# current scope.

sub unimport {
    my ($class, @args) = @_;

    # the only situation in which there is no bindings hash is if this is a "no autobox"
    # that precedes any "use autobox", in which case we don't need to turn autoboxing off as it's
    # not yet been turned on
    return unless ($^H{autobox});

    my $bindings;

    if (@args) {
        $bindings = { %{$^H{autobox}} }; # clone the current bindings hash
        my %args = map { $_ => 1 } @args;

        # expand any virtual type "macros"
        for my $vtype (sort { $ISA{$a}->[0] <=> $ISA{$b}->[0] } keys %ISA) {
            next unless ($args{$vtype});

            # we could delete the types directly from $bindings here, but we may as well pipe them
            # through the option checker below to ensure correctness
            $args{$_} = 1 for (@{$ISA{$vtype}->[1]});

            delete $args{$vtype};
        }

        for my $type (keys %args) {
            # we've handled the virtual types, so we only need to check that this is a valid (real) type
            Carp::confess("unrecognized option: '", (defined $type ? $type : '<undef>'), "'") unless ($TYPES{$type});
            delete $bindings->{$type};
        }
    } else { # turn off autoboxing
        $bindings = {}; # empty hash to trigger full deletion below
    }

    if (%$bindings) {
        _install($bindings);
    } else { # remove all traces of autobox from the current scope
        $^H &= ~0x120000; # unset HINT_LOCALIZE_HH + the additional bit
        delete $^H{autobox};
        delete $^H{autobox_scope};
        delete $^H{autobox_leave}; # triggers the leave handler
    }
}

1;
